From b3b1cbb955425b7f8993f12101ae3f41598a9ca9 Mon Sep 17 00:00:00 2001
From: Ao Xu <ao.xu@amlogic.com>
Date: Mon, 2 Feb 2026 10:56:14 +0800
Subject: [PATCH 174/210] dt-bindings: display: meson: add S4 VPU and HDMI
 compatibles

The S4 VPU splits the HIU registers into clkctrl, pwrctrl and sysctrl
regions and needs the VPU and VAPB clocks. The S4 HDMI transmitter is
compatible with the G12A one apart from secured register access.

Signed-off-by: Ao Xu <ao.xu@amlogic.com>
[chewitt: split from the driver change, and name the power control
 region pwrctrl as the driver and DT do]
Signed-off-by: Christian Hewitt <christianshewitt@gmail.com>
[Armbian: declare the clocks and the reg limits, for dtbs_check]
---
 .../display/amlogic,meson-dw-hdmi.yaml        |  1 +
 .../bindings/display/amlogic,meson-vpu.yaml   | 58 +++++++++++++++++--
 2 files changed, 55 insertions(+), 4 deletions(-)

diff --git a/Documentation/devicetree/bindings/display/amlogic,meson-dw-hdmi.yaml b/Documentation/devicetree/bindings/display/amlogic,meson-dw-hdmi.yaml
index 416fe263ac92..935c83fc4451 100644
--- a/Documentation/devicetree/bindings/display/amlogic,meson-dw-hdmi.yaml
+++ b/Documentation/devicetree/bindings/display/amlogic,meson-dw-hdmi.yaml
@@ -55,6 +55,7 @@ properties:
           - const: amlogic,meson-gx-dw-hdmi
       - enum:
           - amlogic,meson-g12a-dw-hdmi # G12A (S905X2, S905Y2, S905D2)
+          - amlogic,meson-s4-dw-hdmi # S4 (S905Y4)
 
   reg:
     maxItems: 1
diff --git a/Documentation/devicetree/bindings/display/amlogic,meson-vpu.yaml b/Documentation/devicetree/bindings/display/amlogic,meson-vpu.yaml
index 3ae45db85ea7..48327a869a51 100644
--- a/Documentation/devicetree/bindings/display/amlogic,meson-vpu.yaml
+++ b/Documentation/devicetree/bindings/display/amlogic,meson-vpu.yaml
@@ -65,14 +65,21 @@ properties:
           - const: amlogic,meson-gx-vpu
       - enum:
           - amlogic,meson-g12a-vpu # G12A (S905X2, S905Y2, S905D2)
+          - amlogic,meson-s4-vpu # S4 (S905Y4)
 
   reg:
-    maxItems: 2
+    minItems: 2
+    maxItems: 5
 
   reg-names:
-    items:
-      - const: vpu
-      - const: hhi
+    minItems: 2
+    maxItems: 5
+
+  clocks:
+    maxItems: 2
+
+  clock-names:
+    maxItems: 2
 
   interrupts:
     maxItems: 1
@@ -116,6 +123,49 @@ required:
   - "#size-cells"
   - amlogic,canvas
 
+allOf:
+  - if:
+      properties:
+        compatible:
+          contains:
+            const: amlogic,meson-s4-vpu
+    then:
+      properties:
+        reg:
+          items:
+            - description: vcbus registers
+            - description: hhi registers
+            - description: clkctrl registers
+            - description: power control registers
+            - description: sysctrl registers
+        reg-names:
+          items:
+            - const: vpu
+            - const: hhi
+            - const: clkctrl
+            - const: pwrctrl
+            - const: sysctrl
+        clocks:
+          items:
+            - description: vpu clock
+            - description: vapb clock
+        clock-names:
+          items:
+            - const: vpu
+            - const: vapb
+    else:
+      properties:
+        reg:
+          items:
+            - description: vcbus registers
+            - description: hhi registers
+        reg-names:
+          items:
+            - const: vpu
+            - const: hhi
+        clocks: false
+        clock-names: false
+
 additionalProperties: false
 
 examples:
